src/analysis/retail/druid/feral/modules/spells/RipUptimeAndSnapshots.tsx

```diff
@@ -0,0 +1,254 @@
+       *  Clip Duration (but upgrade Snapshot) -> Yellow
+       *  Clip Duration -> Red
+       *  None of the Above -> Green
+       */
+      let value: QualitativePerformance = 'good';
+      if (cast.cpsUsed < MAX_CPS && index !== 0) {
+        value = 'fail';
+      } else if (hasBt && !hasSpec(cast.snapshots, BLOODTALONS_SPEC)) {
+        value = 'fail';
+      } else if (cast.clipped > 0) {
+        value = cast.wasUpgrade ? 'ok' : 'fail';
+      } else if (!hasSpec(cast.snapshots, TIGERS_FURY_SPEC)) {
+        value = 'ok';
+      }
+      // TODO require TF / BT
+
+      const tooltip = (
+        <>
+          @ <strong>{this.owner.formatTimestamp(cast.timestamp)}</strong> targetting{' '}
+          <strong>{cast.targetName || 'unknown'}</strong> using <strong>{cast.cpsUsed} CPs</strong>
+          <br />
+          {!cast.wasNewApplication && (
+            <>
+              Refreshed on target w/ {(cast.remainingOnPrev / 1000).toFixed(1)}s remaining{' '}
+              {cast.clipped > 0 && (
+                <>
+                  <strong>- Clipped {(cast.clipped / 1000).toFixed(1)}s!</strong>
+                </>
+              )}
+              <br />
+            </>
+          )}
+          Snapshots: <strong>{cast.snapshots.map((ss) => ss.name).join(', ')}</strong>
+          <br />
+        </>
+      );
+      return { value, tooltip };
+    });
+
+    return (
+      <SubSection>
+        <p>
+          <b>
+            <SpellLink id={SPELLS.RIP.id} />
+          </b>{' '}
+          is your highest damage-per-energy single target spender. Try to maintain 100% uptime.{' '}
+          {hasPw ? (
+            <>
+              Use <SpellLink id={TALENTS_DRUID.PRIMAL_WRATH_TALENT.id} /> to apply it when you can
+              hit more than one target.
+            </>
+          ) : (
+            <>
+              You can even keep it active on multiple targets, though if a fight will frequently
+              have multiple targets consider speccing for{' '}
+              <SpellLink id={TALENTS_DRUID.PRIMAL_WRATH_TALENT.id} />.
+            </>
+          )}{' '}
+          Don't refresh early, and try to always snapshot <SpellLink id={SPELLS.TIGERS_FURY.id} />
+          {hasBt && (
+            <>
+              {' '}
+              and <SpellLink id={TALENTS_DRUID.BLOODTALONS_TALENT.id} />
+            </>
+          )}
+          .
+        </p>
+        <strong>Rip uptime / snapshots</strong>
+        <small> - Try to get as close to 100% as the encounter allows!</small>
+        {this.subStatistic()}
+        <strong>Rip casts</strong>
+        <small>
+          {' '}
+          - Green is a good cast, Yellow is an ok cast (clipped duration but upgraded snapshot or
+          missing Tigers Fury), Red is a bad cast (clipped duration
+          {hasBt && ' or missing Bloodtalons'}). Mouseover for more details.
+        </small>
+        <PerformanceBoxRow values={castPerfBoxes} />
+      </SubSection>
+    );
+  }
```

Comment:
it looks like a lot of this ended up shared between rip/rake. not identical, so i'm not sure the extent of refactoring that'd be possible. but a note for future work